src/network/address.cpp

```diff
@@ -0,0 +1,280 @@
+	#include <netdb.h>
+	#include <unistd.h>
+	#include <arpa/inet.h>
+	#define LAST_SOCKET_ERR() (errno)
+	typedef int socket_t;
+#endif
+
+/*
+	Address
+*/
+
+Address::Address()
+{
+	memset(&m_address, 0, sizeof(m_address));
+}
+
+Address::Address(u32 address, u16 port)
+{
+	memset(&m_address, 0, sizeof(m_address));
+	setAddress(address);
+	setPort(port);
+}
+
+Address::Address(u8 a, u8 b, u8 c, u8 d, u16 port)
+{
+	memset(&m_address, 0, sizeof(m_address));
+	setAddress(a, b, c, d);
+	setPort(port);
+}
+
+Address::Address(const IPv6AddressBytes *ipv6_bytes, u16 port)
+{
+	memset(&m_address, 0, sizeof(m_address));
+	setAddress(ipv6_bytes);
+	setPort(port);
+}
+
+// Equality (address family, address and port must be equal)
+bool Address::operator==(const Address &address)
+{
+	if (address.m_addr_family != m_addr_family || address.m_port != m_port)
+		return false;
+
+	if (m_addr_family == AF_INET) {
+		return m_address.ipv4.sin_addr.s_addr ==
+		       address.m_address.ipv4.sin_addr.s_addr;
+	}
+
+	if (m_addr_family == AF_INET6) {
+		return memcmp(m_address.ipv6.sin6_addr.s6_addr,
+		              address.m_address.ipv6.sin6_addr.s6_addr, 16) == 0;
+	}
+
+	return false;
+}
+
+bool Address::operator!=(const Address &address)
+{
+	return !(*this == address);
+}
+
+void Address::Resolve(const char *name)
+{
+	if (!name || name[0] == 0) {
+		if (m_addr_family == AF_INET) {
+			setAddress((u32) 0);
+		} else if (m_addr_family == AF_INET6) {
+			setAddress((IPv6AddressBytes*) 0);
+		}
+		return;
+	}
+
+	struct addrinfo *resolved, hints;
+	memset(&hints, 0, sizeof(hints));
+
+	// Setup hints
+	hints.ai_socktype = 0;
+	hints.ai_protocol = 0;
+	hints.ai_flags    = 0;
+	if(g_settings->getBool("enable_ipv6"))
+	{
+		// AF_UNSPEC allows both IPv6 and IPv4 addresses to be returned
+		hints.ai_family = AF_UNSPEC;
+	}
+	else
+	{
+		hints.ai_family = AF_INET;
+	}
+
+	// Do getaddrinfo()
+	int e = getaddrinfo(name, NULL, &hints, &resolved);
+	if (e != 0)
+		throw ResolveError(gai_strerror(e));
+
+	// Copy data
+	if (resolved->ai_family == AF_INET) {
+		struct sockaddr_in *t = (struct sockaddr_in *) resolved->ai_addr;
+		m_addr_family = AF_INET;
+		m_address.ipv4 = *t;
+	} else if(resolved->ai_family == AF_INET6) {
+		struct sockaddr_in6 *t = (struct sockaddr_in6 *) resolved->ai_addr;
+		m_addr_family = AF_INET6;
+		m_address.ipv6 = *t;
+	} else {
+		freeaddrinfo(resolved);
+		throw ResolveError("");
+	}
+	freeaddrinfo(resolved);
+}
+
+// IP address -> textual representation
+std::string Address::serializeString() const
+{
+// windows XP doesnt have inet_ntop, maybe use better func
+#ifdef _WIN32
+	if(m_addr_family == AF_INET)
+	{
+		u8 a, b, c, d;
+		u32 addr;
+		addr = ntohl(m_address.ipv4.sin_addr.s_addr);
+		a = (addr & 0xFF000000) >> 24;
+		b = (addr & 0x00FF0000) >> 16;
+		c = (addr & 0x0000FF00) >> 8;
+		d = (addr & 0x000000FF);
+		return itos(a) + "." + itos(b) + "." + itos(c) + "." + itos(d);
+	}
+	else if(m_addr_family == AF_INET6)
+	{
```

Comment:
Various bracket and spacing style errors. Would you be so nice and make this section look good as well?